src/lib/CMakeLists.txt

```diff
@@ -602,7 +602,7 @@ set(LINE "#ifndef &\\n#define &(...)\\n#endif\\n")
 add_custom_command (
     OUTPUT ${PROBES_FILE}
     DEPENDS ${SOURCES}
-    COMMAND grep -r --include=*.[ch]pp --exclude=probes.hpp --exclude=provider.hpp --exclude=probe_definitions.hpp -h DTRACE_PROBE ${CMAKE_SOURCE_DIR} | grep "^[[:space:]]*[^[:space:]\\/\\*]" | sed "s/[^(]*(\\([^,]*\\), \\([^(,|))]*\\).*/\\1_\\2/" | uniq | xargs -n 1 -I & printf "${LINE}" | sed "1 i\ // This file is auto-generated by cmake\\n" > ${PROBES_FILE}
+    COMMAND echo // This file is auto-generated by cmake > ${PROBES_FILE} && grep -r --include=*.[ch]pp --exclude=probes.hpp --exclude=usdt_provider.hpp --exclude=usdt_probe_definitions.hpp -h DTRACE_PROBE ${CMAKE_SOURCE_DIR} | grep "^[[:space:]]*[^[:space:]\\/\\*]" | sed "s/[^(]*(\\([^,]*\\), \\([^(,|))]*\\).*/\\1_\\2/" | uniq | xargs -n 1 -I & printf "${LINE}" >> ${PROBES_FILE}
```

Comment:
You changed the files ignored by `grep` to  `--exclude=usdt_provider.hpp --exclude=usdt_probe_definitions.hpp` , but those don't exist. Hence, files which are supposed to be ignored by `grep` are no longer excluded.

Reply:
Argh, that shouldn't have made it in. I renamed those files because it would otherwise be somewhat unclear what they do in the build folder. When undoing that, I missed that part, which led to the parallelism problem.

The other problem exists without the `usdt_` prefix too:

```
In file included from /Users/markus/Projekte/Opossum/Git/src/lib/expression/evaluation/expression_evaluator.cpp:30:
In file included from /Users/markus/Projekte/Opossum/Git/src/lib/scheduler/current_scheduler.hpp:7:
In file included from /Users/markus/Projekte/Opossum/Git/src/lib/utils/tracing/probes.hpp:22:
/Users/markus/Projekte/Opossum/Git/build/provider.hpp:20:9: error: 'HYRISE_JOB_END' macro redefined [-Werror,-Wmacro-redefined]
#define HYRISE_JOB_END(arg0, arg1) \
        ^
```

Reply:
But if it is included later, all those `ifndef` will be false. To me, it looks like the file being included in probes.hpp doesn't do anything, so where is it actually used?

Reply:
Probe names will be converted in `probes.hpp`. These converted tokens are supposed to be matched with the symbols defined by dtrace. If this isn't the case (probe is not present in `provider.d`), a compile error will be raised. To avoid this, we need the `probe_definitions.hpp`.

Reply:
> If this isn't the case (probe is not present in provider.d), a compile error will be raised. 

But isn't that what we want? That sounds better than silently replacing the macros with something that does nothing.

Reply:
When developing on Ubuntu one might forget to add the probe to `provider.d`, since this is not required for SystemTap. A linter rule could be a solution, but I see a problem at determining the correct data type. On Linux you don't have the chance to verify if the content in `provider.d` is correct.

Reply:
Well, the current approach obviously does not work, because a couple of probes are not defined in `provider.d`.

How about this for `lint.sh`?

```
exitcode=0
for probe in $(grep -r --include=*.[ch]pp --exclude=probes.hpp --exclude=provider.hpp --exclude=probe_definitions.hpp -h '^\s*DTRACE_PROBE' ${CMAKE_SOURCE_DIR} | sed -E 's/^ *DTRACE_PROBE[0-9]\(HYRISE, *([A-Z_]+).*$/\1/'); do
    grep -i $probe ${CMAKE_SOURCE_DIR}/lib/utils/tracing/provider.d > /dev/null
    if [ $? -ne 0 ]; then
        echo Probe $probe is not defined in provider.h
        exitcode=1
    fi
done
$(exit $exitcode)
```

Currently, this gives me:
```
Probe OPERATOR_STARTED is not defined in provider.
Probe OPERATOR_EXECUTED is not defined in provider.
Probe OPERATOR_TASKS is not defined in provider.
Probe TASKS_PER_STATEMENT is not defined in provider.
Probe SUMMARY is not defined in provider.
Probe CREATE_PIPELINE is not defined in provider.
Probe PIPELINE_CREATION_DONE is not defined in provider.
Probe SQL_PARSING is not defined in provider.
```

Reply:
Thought about this as well. However, if a probe is not defined correctly in provider file (divergent number of arguments or wrong data type of arguments), a compile error still occurs.

Reply:
Which, at some point in the future should be caught by CI. Until then, a compilation error is still better than something simply not working.